sql: limit hint to reduce the first scan chunk

RaduBerinde opened the pull request:
We plumb a "limit hint" down to the scanNode which will set a batch size in the
kvFetcher. The strategy here is to calculate the first batch according to the
limit in an optimistic fashion (assume any filters will pass), and if that is
not sufficient resume using the regular batch size for any subsequent batches.
Addresses #4299.

```
name                          old time/op  new time/op  delta
Scan1000Limit1_Cockroach-4    2.91ms ±43%  0.40ms ±33%  -86.23%  (p=0.000 n=10+10)
Scan1000Limit10_Cockroach-4   2.61ms ±11%  0.40ms ± 5%  -84.60%  (p=0.000 n=9+8)
Scan1000Limit100_Cockroach-4  2.87ms ±29%  0.73ms ± 7%  -74.52%  (p=0.000 n=9+10)
```

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4778)

<!-- Reviewable:end -->

petermattis commented:
LGTM. That's a very nice speedup.

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4778#-:-KBt5Vh-oC6KGj5ycXjq:1542786086)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: limit hint to reduce the first scan chunk

We plumb a "limit hint" down to the scanNode which will set a batch size in the
kvFetcher. The strategy here is to calculate the first batch according to the
limit in an optimistic fashion (assume any filters will pass), and if that is
not sufficient resume using the regular batch size for any subsequent batches.
Addresses #4299.

```
name                          old time/op  new time/op  delta
Scan1000Limit1_Cockroach-4    2.91ms ±43%  0.40ms ±33%  -86.23%  (p=0.000 n=10+10)
Scan1000Limit10_Cockroach-4   2.61ms ±11%  0.40ms ± 5%  -84.60%  (p=0.000 n=9+8)
Scan1000Limit100_Cockroach-4  2.87ms ±29%  0.73ms ± 7%  -74.52%  (p=0.000 n=9+10)
```
